gdx/src/com/badlogic/gdx/graphics/g2d/TextureArraySpriteBatch.java

```diff
@@ -0,0 +1,1415 @@
+	 * The defaultShader specifies the shader to use. Note that the names for uniforms for this default shader are different than
+	 * the ones expect for shaders set with {@link #setShader(ShaderProgram)}. See {@link#createDefaultShader()}.
+	 * @param size The max number of sprites in a single batch. Max of 8191.
+	 * @param defaultShader The default shader to use. This is not owned by the TextureArraySpriteBatch and must be disposed
+	 *           separately.
+	 * @throws IllegalStateException If the device does not support texture arrays. */
+	public TextureArraySpriteBatch (int size, ShaderProgram defaultShader) throws IllegalStateException {
+
+		// 32767 is max vertex index, so 32767 / 4 vertices per sprite = 8191 sprites max.
+		if (size > 8191) throw new IllegalArgumentException("Can't have more than 8191 sprites per batch: " + size);
+
+		// Query the number of available texture units and decide on a safe number of texture units to use
+		IntBuffer texUnitsQueryBuffer = BufferUtils.newIntBuffer(32);
+
+		Gdx.gl.glGetIntegerv(GL20.GL_MAX_TEXTURE_IMAGE_UNITS, texUnitsQueryBuffer);
+
+		maxTextureUnits = texUnitsQueryBuffer.get();
+
+		// Some OpenGL drivers (I'm looking at you, Intel!) do not report the right values,
+		// so we take caution and test it first, reducing the number of slots if needed.
+		if (defaultShader == null) {
+
+			// Will try to create a shader with a lower amount of texture units if creation fails.
+			while (maxTextureUnits > 0) {
+
+				try {
+
+					shader = createDefaultShader(maxTextureUnits);
+
+					break;
+
+				} catch (Exception e) {
+					maxTextureUnits /= 2;
+				}
+			}
+
+			if (maxTextureUnits == 0) {
+				throw new IllegalStateException("Texture Arrays are not supported on this device.");
+			}
+
+			ownsShader = true;
+
+		} else {
+			shader = defaultShader;
+			ownsShader = false;
+		}
+
+		usedTextures = new Texture[maxTextureUnits];
+		usedTexturesLFU = new int[maxTextureUnits];
+
+		// This contains the numbers 0 ... maxTextureUnits - 1. We send these to the shader as an uniform.
+		textureUnitIndicesBuffer = BufferUtils.newIntBuffer(maxTextureUnits);
+		for (int i = 0; i < maxTextureUnits; i++) {
+			textureUnitIndicesBuffer.put(i);
+		}
+		textureUnitIndicesBuffer.flip();
+
+		VertexDataType vertexDataType = (Gdx.gl30 != null) ? VertexDataType.VertexBufferObjectWithVAO : VertexDataType.VertexArray;
+
+		// The vertex data is extended with one float for the texture index.
+		mesh = new Mesh(vertexDataType, false, size * 4, size * 6,
+			new VertexAttribute(Usage.Position, 2, ShaderProgram.POSITION_ATTRIBUTE),
+			new VertexAttribute(Usage.ColorPacked, 4, ShaderProgram.COLOR_ATTRIBUTE),
+			new VertexAttribute(Usage.TextureCoordinates, 2, ShaderProgram.TEXCOORD_ATTRIBUTE + "0"),
+			new VertexAttribute(Usage.Generic, 1, "texture_index"));
+
+		projectionMatrix.setToOrtho2D(0, 0, Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
+
+		vertices = new float[size * (Sprite.SPRITE_SIZE + 4)];

```

Comment:
I think we should be consistent on whether we use `Sprite.SPRITE_SIZE` or `spriteFloatSize`

Reply:
Good call, gonna make this consistent tomorrow!